File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/InboxFragment.java
package com.habitrpg.android.habitica.ui.fragments.social;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.SocialRepository;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.models.social.ChatMessage;
import com.habitrpg.android.habitica.models.user.User;
import com.habitrpg.android.habitica.modules.AppModule;
import com.habitrpg.android.habitica.prefs.scanner.IntentIntegrator;
import com.habitrpg.android.habitica.prefs.scanner.IntentResult;
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment;
import com.habitrpg.android.habitica.ui.helpers.UiUtils;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.Sort;

public class InboxFragment extends BaseMainFragment
        implements SwipeRefreshLayout.OnRefreshListener, View.OnClickListener {

    @Inject
    SocialRepository socialRepository;
    @Inject
    @Named(AppModule.NAMED_USER_ID)
    String userId;

    @BindView(R.id.inbox_messages)
    LinearLayout inboxMessagesListView;

    @BindView(R.id.inbox_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;

    private View chooseRecipientDialogView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        this.socialRepository.markPrivateMessagesRead(user)
                .subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());

        View v = inflater.inflate(R.layout.fragment_inbox, container, false);
        unbinder = ButterKnife.bind(this, v);

        swipeRefreshLayout.setOnRefreshListener(this);

        loadMessages();

        return v;
    }

    private void loadMessages() {
        if (user != null && user.isManaged()) {
            user.getInbox().getMessages().where()
                    .distinct("uuid")
                    .sort("timestamp", Sort.DESCENDING)
                    .asObservable()
                    .first()
                    .subscribe(this::setInboxMessages, RxErrorHandler.handleEmptyError());
        }
    }

    @Override
    public void onDestroy() {
        socialRepository.close();
        super.onDestroy();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (this.activity != null) {
            this.activity.getMenuInflater().inflate(R.menu.inbox, menu);
        }
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.send_message:
                openNewMessageDialog();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void openNewMessageDialog() {
        assert this.activity != null;
        this.chooseRecipientDialogView = this.activity.getLayoutInflater().inflate(R.layout.dialog_choose_message_recipient, null);

        Button scaneQRCodeButton = (Button) chooseRecipientDialogView.findViewById(R.id.scanQRCodeButton);

        AlertDialog alert = new AlertDialog.Builder(this.activity)
                .setTitle(getString(R.string.choose_recipient_title))
                .setPositiveButton(getString(R.string.action_continue), (dialog, which) -> {
                    EditText uuidEditText = (EditText) chooseRecipientDialogView.findViewById(R.id.uuidEditText);
                    openInboxMessages(uuidEditText.getText().toString(), "");
                })
                .setNeutralButton(getString(R.string.action_cancel), (dialog, which) -> {
                    UiUtils.dismissKeyboard(this.activity);
                    dialog.cancel();
                })
                .create();
        scaneQRCodeButton.setOnClickListener((View v) -> {
            IntentIntegrator scanIntegrator = new IntentIntegrator(getActivity());
            scanIntegrator.initiateScan(this);
        });
        alert.setView(chooseRecipientDialogView);
        alert.show();
    }

    @Override
    public void injectFragment(AppComponent component) {
        component.inject(this);
    }

    @Override
    public void onRefresh() {
        swipeRefreshLayout.setRefreshing(true);
        this.userRepository.retrieveUser(true)
                .subscribe(this::onUserReceived, RxErrorHandler.handleEmptyError());
    }

    public void setInboxMessages(List<ChatMessage> messages) {
        if (this.inboxMessagesListView == null) {
            return;
        }

        this.inboxMessagesListView.removeAllViewsInLayout();

        LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        if (messages.size() > 0) {
            for (ChatMessage message : messages) {
                TextView entry = (TextView) inflater.inflate(R.layout.plain_list_item, this.inboxMessagesListView, false);
                entry.setText(message.user);
                entry.setTag(message.uuid);
                entry.setOnClickListener(this);
                this.inboxMessagesListView.addView(entry);
            }
        } else {
            TextView tv = new TextView(getContext());
            tv.setText(R.string.empty_inbox);
        }
    }

    @Override
    public void onClick(View v) {
        TextView entry = (TextView) v;
        String replyToUserName = entry.getText().toString();
        openInboxMessages(entry.getTag().toString(), replyToUserName);
    }

    private void openInboxMessages(String userID, String username) {
        InboxMessageListFragment inboxMessageListFragment = new InboxMessageListFragment();
        inboxMessageListFragment.setReceivingUser(username, userID);
        if (this.activity != null) {
            this.activity.displayFragment(inboxMessageListFragment);
        }
    }

    public void onUserReceived(User user) {
        this.user = user;
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        IntentResult scanningResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);

        if (scanningResult != null && scanningResult.getContents() != null) {
            if (this.chooseRecipientDialogView != null) {
                EditText uuidEditText = (EditText) this.chooseRecipientDialogView.findViewById(R.id.uuidEditText);
                String qrCodeUrl = scanningResult.getContents();
                Uri uri = Uri.parse(qrCodeUrl);
                if (uri == null || uri.getPathSegments().size() < 3) {
                    return;
                }
                String userID = uri.getPathSegments().get(2);
                uuidEditText.setText(userID);
            }
        }
    }

    @Override
    public String customTitle() {
        if (!isAdded()) {
            return "";
        }
        return getString(R.string.sidebar_inbox);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/InboxFragment.kt
package com.habitrpg.android.habitica.ui.fragments.social

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.components.UserComponent
import com.habitrpg.android.habitica.data.SocialRepository
import com.habitrpg.android.habitica.helpers.AppConfigManager
import com.habitrpg.android.habitica.helpers.MainNavigationController
import com.habitrpg.android.habitica.helpers.RxErrorHandler
import com.habitrpg.android.habitica.models.social.ChatMessage
import com.habitrpg.android.habitica.modules.AppModule
import com.habitrpg.android.habitica.ui.AvatarView
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment
import com.habitrpg.android.habitica.ui.helpers.dismissKeyboard
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaAlertDialog
import com.habitrpg.android.habitica.ui.views.social.UsernameLabel
import io.reactivex.functions.Consumer
import io.realm.RealmResults
import kotlinx.android.synthetic.main.fragment_inbox.*
import javax.inject.Inject
import javax.inject.Named

class InboxFragment : BaseMainFragment(), androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener, View.OnClickListener {

    @Inject
    lateinit var socialRepository: SocialRepository
    @field:[Inject Named(AppModule.NAMED_USER_ID)]
    lateinit var userId: String
    @Inject
    lateinit var configManager: AppConfigManager

    private var chooseRecipientDialogView: View? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        this.hidesToolbar = true
        super.onCreateView(inflater, container, savedInstanceState)

        compositeSubscription.add(this.socialRepository.markPrivateMessagesRead(user).subscribe(Consumer { }, RxErrorHandler.handleEmptyError()))

        return inflater.inflate(R.layout.fragment_inbox, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        inbox_refresh_layout?.setOnRefreshListener(this)

        loadMessages()
    }

    private fun loadMessages() {
        compositeSubscription.add(socialRepository.getInboxOverviewList().subscribe(Consumer<RealmResults<ChatMessage>> {
            setInboxMessages(it)
        }, RxErrorHandler.handleEmptyError()))
    }

    override fun onDestroy() {
        socialRepository.close()
        super.onDestroy()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        when (id) {
            R.id.send_message -> {
                openNewMessageDialog()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun openNewMessageDialog() {
        assert(this.activity != null)
        this.chooseRecipientDialogView = this.activity?.layoutInflater?.inflate(R.layout.dialog_choose_message_recipient, null)

        this.activity?.let { thisActivity ->
            val alert = HabiticaAlertDialog(thisActivity)
            alert.setTitle(getString(R.string.choose_recipient_title))
            alert.addButton(getString(R.string.action_continue), true) { _, _ ->
                    val uuidEditText = chooseRecipientDialogView?.findViewById<View>(R.id.uuidEditText) as? EditText
                    openInboxMessages(uuidEditText?.text?.toString() ?: "", "")
                }
            alert.addButton(getString(R.string.action_cancel), false) { dialog, _ ->
                    thisActivity.dismissKeyboard()
                }
            alert.setAdditionalContentView(chooseRecipientDialogView)
            alert.show()
        }

    }

    override fun injectFragment(component: UserComponent) {
        component.inject(this)
    }

    override fun onRefresh() {
        inbox_refresh_layout.isRefreshing = true
        compositeSubscription.add(this.socialRepository.retrieveInboxMessages()
                .subscribe(Consumer<List<ChatMessage>> {
                    inbox_refresh_layout.isRefreshing = false
                }, RxErrorHandler.handleEmptyError()))
    }

    private fun setInboxMessages(messages: RealmResults<ChatMessage>) {
        if (inbox_messages == null) {
            return
        }

        inbox_messages.removeAllViewsInLayout()

        val inflater = context?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as? LayoutInflater
        if (messages.isNotEmpty()) {
            for (message in messages) {
                val entry = inflater?.inflate(R.layout.item_inbox_overview, inbox_messages, false)
                val avatarView = entry?.findViewById(R.id.avatar_view) as? AvatarView
                //message.userStyles?.let { avatarView?.setAvatar(it) }
                avatarView?.visibility = View.GONE
                val displayNameTextView = entry?.findViewById(R.id.display_name_textview) as? UsernameLabel
                displayNameTextView?.username = message.user
                displayNameTextView?.tier = message.contributor?.level ?: 0
                val timestampTextView = entry?.findViewById(R.id.timestamp_textview) as? TextView
                timestampTextView?.text = message.getAgoString(resources)
                val usernameTextView = entry?.findViewById(R.id.username_textview) as? TextView
                if (message.username != null) {
                    usernameTextView?.text = message.formattedUsername
                    usernameTextView?.visibility = View.VISIBLE
                } else {
                    usernameTextView?.visibility = View.GONE
                }
                val messageTextView = entry?.findViewById(R.id.message_textview) as? TextView
                messageTextView?.text = message.text
                entry?.tag = message.uuid
                entry?.setOnClickListener(this)
                inbox_messages.addView(entry)
            }
        } else {
            val tv = TextView(context)
            tv.setText(R.string.empty_inbox)
        }
    }

    override fun onClick(v: View) {
        val displaynameView = v.findViewById(R.id.display_name_textview) as? UsernameLabel
        val replyToUserName = displaynameView?.username ?: ""
        openInboxMessages(v.tag.toString(), replyToUserName)
    }

    private fun openInboxMessages(userID: String, username: String) {
        MainNavigationController.navigate(InboxFragmentDirections.openInboxDetail(userID, username))
    }

}
